# Analysing user data to maximise in-app ads outreach

For this project, we take on the role of a data analyst at an app development company, which builds mobile apps for both Android and iOS operating systems, and publishes the apps on both Google Play Store and Apple Apps Store. The company is looking to develop a new app that is free to download and has its revenus driven via in-app ads.

This project aims to analyse exisiting applications in both markets to construct user profiles in hope to provide developers detailed insights to better customise app type and features to encapsulate the largest audience and hence increase in-app advertisment impact to drive revenue.

To achieve this goal, we will analyse apps that are currently avaliable on the markets to understand the features of popular apps. We will download data sets containing the information of current apps on both platforms.

As there are approximately 2 million apps in the App Store and 2.1 million apps in Google Play as of September 2018, we have chosen to use samples of the full apps data file because the complete file is too large and costly to analyse. Before exploring the data, we will first open and read in two datasets that contains samples of apps that are on the App Store and Google Play Markets:

In [1]:
## Opening and reading the dataset ##
import csv # For reader() function

## Android apps data ##
opened_file = open(r'C:\Users\tsa19\Dataquest Projects\Datasets\googleplaystore.csv', encoding="utf8")
# Note: encountered SyntaxError: (unicode error) when did not convert file path to raw string by adding r"text".
# Note: encountered UnicodeDecodeError, specify encoding method to fix.
read_file = csv.reader(opened_file)
android_full = list(read_file)

## Apple apps data ##
opened_file = open(r'C:\Users\tsa19\Dataquest Projects\Datasets\AppleStore.csv', encoding="utf8")
read_file = csv.reader(opened_file)
apple_full = list(read_file)

In [2]:
## Defining explore dataset function to start understanding the datasets ##
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [3]:
## Isolating header rows ##
android_header = android_full[0]
android = android_full[1:]

apple_header = apple_full[0]
apple = apple_full[1:]

## Explore both datasets with explore data function ##
explore_data(android, 0, 5, rows_and_columns=True)
explore_data(apple, 0, 5, rows_and_columns=True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 10841
Number of columns: 13
['284882215', 'Facebook', '38987980

In [4]:
## Printing column names of both dataset and identifying which variables are useful to the project. ##
print("These are the column names for dataset android: ")
print(android_header)

print('\n')

print("These are the column names for dataset apple: ")
print(apple_header)

These are the column names for dataset android: 
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


These are the column names for dataset apple: 
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


## Explaination for dataset 'googleplaystore.csv' and 'AppleStore.csv'

These two data sets are sample data sets that were drawn randomly from the apps listed on Google Play Store and Apple's App Store respectively.
They both include information such as apps name, user rating, prices, genre and more.
The source of both datasets are found on Kaggle.

Now, we will identify the variables that could possibly be of interest to the project.
For googleplaystore.csv, they are: 'App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type' and 'Price'.

You can download and find the documentation for the **Google Play Store dataset** here: [Link](https://www.kaggle.com/lava18/google-play-store-apps)

| **Column Name** | **Description** | **Data Type** |
|:---|:---|:---|
| "App" | Application name | string |
| "Category" | App category | string |
| "Rating" | Overall user rating of the app (as when scraped) | decimal |
| "Reviews" | Number of user reviews for the app (as when scraped) | integer |
| "Size" | Size of the app (as when scraped) | string |
| "Installs" | Number of user downloads/installs for the app (as when scraped) | string |
| "Type" | Paid or Free | string |
| "Price" | Price of the app (as when scraped) | string |

*Side note: link for generating tables in markdown cells can be found here:* [Link](https://www.markdownguide.org/extended-syntax/#tables)

For AppleStore.csv, relevant columns are: 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'user_rating' and 'prime_genre'.

Download path and documentation for **Apple Apps Store dataset** can be found here: [Link](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

| Column Name | Description | Data Type |
|:---|:---|:---|
| "track_name" | Application name | string |
| "size_bytes" | Size (in Bytes) | integer |
| "Currency" | Currency Type | string |
| "Price" | Price amount | decimal |
| "rating_count_tot" | User Rating counts (for all version) | integer |
| "user_rating" | Average User Rating value (for all version) | decimal |
| "prime_genre" | Primary Genre | string |

# Data preparation and cleaning

Now that we have identified the variables that are needed to our project aim, let us prepare the data as to obtain accurate analysis results.

This process is called *data cleaning*, and it is usually the task what data scientists spend the most time on.
The aim for this step is to:
* Detect inaccurate data, and correct or remove it.
* Detect duplicate data, and remove the duplicates.

Also, we will impose two more conditions when preparing the dataset because the project outline has noted that we are interested in building *free* apps towards an *English-speaking audience*, hence we will:
* Remove non-English apps like 爱奇艺PPS -《欢乐颂2》电视剧热播.
* Remove apps that aren't free.

## Removing wrong data

From the discussion from the Kaggle website where this dataset is sourced from, we see a [thread of discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) mentioning this error where the data seemed to have skipped the Category data, and hence is one data short.
Luckily there is only one app that have this issue, and so we decide to correct this error by removing row index 10472.

In [5]:
## Identifying error in data entry length ##

for row in android:
    if len(row) == len(android_header):
        continue
    elif len(row) != len(android_header):
        print(row)
        print("Incorrect row index: ", android.index(row)) # header row is removed in this case

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
Incorrect row index:  10472


In [6]:
## Removing data for this app ##
del android[10472] # take care not to run this code more than once

## Removing Duplicated Entries: Part One

Next, to ensure the accuracy of our analysis, we will investigate the data set to see if there are only one entry per app, and select a method to clean the data if not.

In [7]:
## Finding duplicated data ##
unique_apps = []
duplicated_apps = []

for row in android:
    name = row[0]
    if name in unique_apps:
        duplicated_apps.append(name)
    if name not in unique_apps:
        unique_apps.append(name)
        
print("There are ", len(duplicated_apps), " duplicated apps in the android data set.")
print('\n')

print("Examples of duplicated apps:")
for row in android:
    name = row[0]
    if name == duplicated_apps[0]:
        print(row)

There are  1181  duplicated apps in the android data set.


Examples of duplicated apps:
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80804', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']


Above, we have identified that there are 1,181 duplicated entries in the android data set, but randomly choosing one entry and removing all other duplicates is really a good way to ensure that we retain the best data in our data set.

One criterion for selection could be to keep the entry with the most review, as it will mean that it is the latest data.

Another criterion is to keep the entry with the most recent Last Updated date and/or Current Ver to keep the most recent data.

If all information between entries are the same, then we will select an entry at random and drop the rest.

## Removing Duplicated Entries: Part Two

After setting out the criterion which to select duplicated data from, we will proceed to cleaning the data set.

We will first create a dictionary that maps the unique app names to the largest number of reviews that the app has. After which, we will use this dictionary to extract the entries that have the maximum review number. Here, we will need to first check if data for this app had been extracted already, because there are instances where max review numbers are the same across multiple entries. At this point, we will just extract the first entry that have the max review number as mapped in the dictionary and drop the others.

In [8]:
## Creating dictionary to select most reviewed data entries ##
print("Expected data set length:", len(android) - len(duplicated_apps))
reviews_max = {}

for row in android:
    name = row[0]
    n_reviews = float(row[3])
    if name not in reviews_max:
        reviews_max[name] = n_reviews
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    else:
        pass

print("Dictionary length:", len(reviews_max))

Expected data set length: 9659
Dictionary length: 9659


In [9]:
## Using dictionary to remove duplicated data ##
android_clean = []
already_added = []

for row in android:
    name = row[0]
    n_reviews = float(row[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(row)
        already_added.append(name)

if len(android_clean) == len(reviews_max):
    print("We have successfully removed duplicated data and reduce data entry to", len(android_clean), ".")
else:
    print("Data set still needs to be further cleaned.")

We have successfully removed duplicated data and reduce data entry to 9659 .


## Removing Non-English apps: Part One

We now move to removing non-English apps from our data set.

For this step, we exploit the fact that every string character have an unique integer that represents the Unicode point of that character. For characters that appears in English, according to ASCII (American Standard Code for Information Interchange), this integer is in the range of 0 to 127. Hence if we check the name of the app and characters within the name returns with a integer that is greater than 127, then the app most likely have a non-English name.

We will define a function that help us identify if the app name is detected as English or non-English.

In [10]:
## Defining function to identify if app have English name ##
def english_app_name(string):
    for element in string:
        if ord(element) > 127:
            return False
        else:
            return True
        
test_app_names = ['Instagram', '爱奇艺PPS -《欢乐颂2》电视剧热播', 'Docs To Go™ Free Office Suite', 'Instachat 😜']

for name in test_app_names:
    print(name, "has an English app name:", english_app_name(name))

Instagram has an English app name: True
爱奇艺PPS -《欢乐颂2》电视剧热播 has an English app name: False
Docs To Go™ Free Office Suite has an English app name: True
Instachat 😜 has an English app name: True


Because some English named apps have uncommon characters included, so they will be rejected using the current fuction. If we choose to filter our data set using this fuction we will be loosing large quantity of data. Hence, to minimize the impact of data loss, we will modify the function to only filter out apps with more than three characters that are outside of the ASCII range.

This method is not entirely accurate, however, it is effective enough for the cause of our project.

In [11]:
## Modifying fuction definition to identify if app have English name ##
def english_app_name(string):
    non_eng_count = 0
    for element in string:
        if ord(element) > 127:
            non_eng_count += 1
        else:
            pass
    if non_eng_count > 3:
        return False
    else:
        return True
        
test_app_names = ['Instagram', '爱奇艺PPS -《欢乐颂2》电视剧热播', 'Docs To Go™ Free Office Suite', 'Instachat 😜']

for name in test_app_names:
    print(name, "has an English app name:", english_app_name(name))

Instagram has an English app name: True
爱奇艺PPS -《欢乐颂2》电视剧热播 has an English app name: False
Docs To Go™ Free Office Suite has an English app name: True
Instachat 😜 has an English app name: True


In [12]:
## Using function to filter out non-English named apps ##
android_eng_clean = []

for row in android_clean:
    name = row[0]
    if english_app_name(name) == True:
        android_eng_clean.append(row)

print("Dropped non-Enlish Android apps:", len(android_clean) - len(android_eng_clean))
print("Android data set length:", len(android_eng_clean))

apple_eng_clean = []

for row in apple:
    name = row[1]
    if english_app_name(name) == True:
        apple_eng_clean.append(row)
        
print("Dropped non-Enlish Apple apps:", len(apple) - len(apple_eng_clean))
print("Apple data set length:", len(apple_eng_clean))

Dropped non-Enlish Android apps: 45
Android data set length: 9614
Dropped non-Enlish Apple apps: 1014
Apple data set length: 6183


## Isolating free apps

As we are only interested in analysing data from apps that are free to download in both markets, we are going to isolate these data from the rest of the data set.

In [13]:
## Isolating the free apps ##
android_free = []

for row in android_eng_clean:
    free_var = row[6]
    if free_var == 'Free':
        android_free.append(row)
        
print("Dropped non-free Android apps:", len(android_eng_clean) - len(android_free))
print("Android data set final length:", len(android_free))

apple_free = []

for row in apple_eng_clean:
    free_var = row[4]
    if float(free_var) == 0:
        apple_free.append(row)
        
print("Dropped non-free Apple apps:", len(apple_eng_clean) - len(apple_free))
print("Apple data set final length:", len(apple_free))

Dropped non-free Android apps: 751
Android data set final length: 8863
Dropped non-free Apple apps: 2961
Apple data set final length: 3222


## Most common apps by genre

As described in introduction, our project aim is to identify the kind of app that will attract the most users because the number of users strongly correlates to the revenue generated by said app.

In order to minimise business risks and overheads, we will adopt the following strategy:
1. Build minimal android version of app, and launch it on Google Play.
2. Develop app further if feedbacks were positive.
3. If app is profitable in six months, develop and launch iOS version on App Store.

Seeing that our end goal is to add our app to both markets, we need to determine the traits that makes an app successful across different markets.

We will first start analysing the data by genre, and the variables choosen for this investigation will be:
* Category and Genre for android data set
* prime_genre for apple data set

In [14]:
## Defining function to generate frequency tables ##
def freq_table(dataset, index):
    table = {}
    for row in dataset:
        key = row[index]
        if key in table:
            table[key] += 1
        else:
            table[key] = 1
            
    for key in table:
        table[key] = table[key] / len(dataset) * 100 # transform frequency in to percentatge
        
    return table

## Defining a function to sort frequency table in descending order ##
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [15]:
## Using functions to list genre by popularity ##
print("Frequency table by Category for Google Play data set:")
display_table(android_free, 1)
print('\n')
print("Frequency table by Genre for Google Play data set:")
display_table(android_free, 9)
print('\n')
print("Frequency table by prime_genre for App Store data set:")
display_table(apple_free, 11)

Frequency table by Category for Google Play data set:
FAMILY : 18.898792733837304
GAME : 9.725826469592688
TOOLS : 8.462146000225657
BUSINESS : 4.592124562789123
LIFESTYLE : 3.9038700214374367
PRODUCTIVITY : 3.8925871601038025
FINANCE : 3.7007785174320205
MEDICAL : 3.5315355974275078
SPORTS : 3.396141261423897
PERSONALIZATION : 3.317161232088458
COMMUNICATION : 3.2381812027530184
HEALTH_AND_FITNESS : 3.0802211440821394
PHOTOGRAPHY : 2.944826808078529
NEWS_AND_MAGAZINES : 2.798149610741284
SOCIAL : 2.6627552747376737
TRAVEL_AND_LOCAL : 2.335552296062281
SHOPPING : 2.245289405393208
BOOKS_AND_REFERENCE : 2.1437436533904997
DATING : 1.8616721200496444
VIDEO_PLAYERS : 1.7939749520478394
MAPS_AND_NAVIGATION : 1.399074805370642
FOOD_AND_DRINK : 1.241114746699763
EDUCATION : 1.1621347173643235
ENTERTAINMENT : 0.9590432133589079
LIBRARIES_AND_DEMO : 0.9364774906916393
AUTO_AND_VEHICLES : 0.9251946293580051
HOUSE_AND_HOME : 0.8236488773552973
WEATHER : 0.8010831546880289
EVENTS : 0.710820264018

## Analysing app distribution by genre

We have now created frequency tables which displays results for Category and Genre columns for Google Play data set and prime_genre for App Store data set. The frequencies have been transformed in to percentages and sorted in descending order for ease of analysis.

Understanding the results from the freqency table generated for the prime_genre column for the App Store data set, we find that the most populated genre is by Games with 58.16%, followed by Entertainment at 7.88%. The genres after the top two all have frequencies that are less than 5%, showing that the market is being dominated by apps geared towards entertainment instead of practicle purposes.

One possible reason could be that entertainment apps, and gaming apps in particular, are easier to develop, hence are released on to the market in troves. However, our analysis thus far does not account for the number of users each genre of app have, hence if the large quantity of gaming apps only attracts small number of users each, then developing a gaming app will not be compatible with our revenue generating method. We would require further investigation into the spread of user download frequencies to draw more accurate conclusion.

Another point to be made is that the analysis is conducted on an English free app data set, hence it could be that different genre of games adopt different business strategies, where gaming and other entertainment apps aims to minimize friction to encourage the largest number of user installation by removing inital price barrier, then set up in-app payment features or adopts the ad driven revenue strategy that our company sets out to follow. Productivity apps, on the otherhand, most likely adopts different pricing strategies and are not free apps as we know that out of approximately 6000 English apps in our App Store data set, roughly half are paid apps and therefore dropped from the analysis data set.

For genres in Google Play market, the most popular genre by Category column is Family (18.90%) and followed by Game (9.73%); results by Genre column are Tools (8.45%) and Entertainment (6.07%). The spread of frequency of the Google Play Store data set it much less extreme than that of the App Store's, but the top genre is still twice as populated as the second. The pattern between entertainment and productivity apps is less apparent for the android market, but if we interpret Category as the primary genre, and the Genre column as the secondary  genres, then we would find that android apps are more evenly spread across genres, though it would appear entertainment apps are slightly more popluar.

It is important to reiterate that the analysis above are conducted on English free apps for both markets, hence the results might not be applicable to the markets as a whole. But with our current project specification,

In order to suggest an app profile for our specification, we would need to investigate the average number of user installs.

## Most poplular apps by genre on the App Store

Because we have no variable for user installation for the App Store data set, we will use user_rating, which records the count of user ratings submitted for each app as a proxy.

We will stratify these results by genre.

In [16]:
## Finding most popular app by genre ##
app_store_genre = freq_table(apple_free, 11)
avg_user_num_map = {}

for genre in app_store_genre:
    total = 0
    len_genre = 0
    for row in apple_free:
        genre_app = row[11]
        if genre_app == genre:
            total = total + float(row[5]) # Chosen to use rating_count_tot
            len_genre += 1
    avg_user_num = round(total / len_genre, 2)
    avg_user_num_map[genre] = avg_user_num
    
table_sorted = sorted(avg_user_num_map.items(), key=lambda kv: kv[1], reverse=True)

for entry in table_sorted:
       print(entry[1], ':', entry[0])

86090.33 : Navigation
74942.11 : Reference
71548.35 : Social Networking
57326.53 : Music
52279.89 : Weather
39758.5 : Book
33333.92 : Food & Drink
31467.94 : Finance
28441.54 : Photo & Video
28243.8 : Travel
26919.69 : Shopping
23298.02 : Health & Fitness
23008.9 : Sports
22788.67 : Games
21248.02 : News
21028.41 : Productivity
18684.46 : Utilities
16485.76 : Lifestyle
14029.83 : Entertainment
7491.12 : Business
7003.98 : Education
4004.0 : Catalogs
612.0 : Medical


We can see that the most popular app genres are Navigation, Reference and Social Networking. These apps require large amount of user generated content to optimize its performance and thus users have been monopolised by the few apps operating within these spheres.

We can also observe that our previous prediction about user number being diluted by quantity of apps avaliable is well supported as average number of users per Gaming app ranks pretty low on the list.

One recommened profile would be a Photo & Video app, this genre is not overly crowed as Games, but yields a pretty high number of user downloads on average. Hence could be a good fit for our requirement.

In [17]:
## Finding most popular app by genre ##
cat_genre = freq_table(android_free, 1)
avg_user_num_map = {}

for category in cat_genre:
    total = 0
    len_genre = 0
    for row in android_free:
        category_app = row[1]
        if category_app == category:
            installs = row[5]
            installs = installs.replace("+", "")
            installs = installs.replace(",", "")
            total = total + float(installs)
            len_genre += 1
    avg_user_num = round(total / len_genre, 2)
    avg_user_num_map[category] = avg_user_num
    
table_sorted = sorted(avg_user_num_map.items(), key=lambda kv: kv[1], reverse=True)

for entry in table_sorted:
       print(entry[1], ':', entry[0])

38456119.17 : COMMUNICATION
24727872.45 : VIDEO_PLAYERS
23253652.13 : SOCIAL
17840110.4 : PHOTOGRAPHY
16787331.34 : PRODUCTIVITY
15588015.6 : GAME
13984077.71 : TRAVEL_AND_LOCAL
11640705.88 : ENTERTAINMENT
10801391.3 : TOOLS
9549178.47 : NEWS_AND_MAGAZINES
8767811.89 : BOOKS_AND_REFERENCE
7036877.31 : SHOPPING
5201482.61 : PERSONALIZATION
5074486.2 : WEATHER
4188821.99 : HEALTH_AND_FITNESS
4056941.77 : MAPS_AND_NAVIGATION
3697848.17 : FAMILY
3638640.14 : SPORTS
1986335.09 : ART_AND_DESIGN
1924897.74 : FOOD_AND_DRINK
1833495.15 : EDUCATION
1712290.15 : BUSINESS
1437816.27 : LIFESTYLE
1387692.48 : FINANCE
1331540.56 : HOUSE_AND_HOME
854028.83 : DATING
817657.27 : COMICS
647317.82 : AUTO_AND_VEHICLES
638503.73 : LIBRARIES_AND_DEMO
542603.62 : PARENTING
513151.89 : BEAUTY
253542.22 : EVENTS
120550.62 : MEDICAL


The genres with the highest installs per app on the Google Play Market are communication, video players and social.

We see that average download numbers are much higher in the Google Play Market than in the App Store.

Since our aim is to identify the profile that is most likely to maximised profit generations across both platforms, I would suggest for the company to develop an application that would fit in the Photo genre.

Note to self: Fine tune the project following this [style guide](https://www.dataquest.io/blog/data-science-project-style-guide/)